kategory-core/src/main/kotlin/kategory/data/Weak.kt

```diff
@@ -0,0 +1,67 @@
+package kategory
+
+import java.lang.ref.WeakReference
+
+/**
+ * Represents an object that can stop existing when no references to it exists. Backed by
+ * a [WeakReference] instance.
+ */
+@higherkind
+@deriving(
+    Functor::class,
+    Applicative::class,
+    Monad::class,
+    Foldable::class,
+    Traverse::class,
+    TraverseFilter::class,
+    MonadFilter::class)
+data class Weak<out A>(val provider: () -> A?) : WeakKind<A> {
+
+    companion object {
+
+        private val EMPTY: Weak<Nothing> = Weak { null }
+
+        @Suppress("UNCHECKED_CAST")
+        fun <B> emptyWeak(): Weak<B> = EMPTY
+
+        operator fun <A> invoke(a: A): Weak<A> {
+            val reference = WeakReference(a)
+            return Weak { reference.get() }
+        }
+    }
+
+    inline fun <B> fold(fn: () -> B, f: (A) -> B): B = provider().let {
+        when (it) {
+            null -> fn()
+            else -> f(it)
+        }
+    }
+
+    inline fun <B> map(crossinline f: (A) -> B): Weak<B> = fold({ emptyWeak() }, { f(it).weak() })
+
+    inline fun <B> flatMap(crossinline f: (A) -> Weak<B>): Weak<B> = fold({ emptyWeak() }, { a -> f(a) })
```

Comment:
`crossinline f: (A) -> WeakKind<B>` and `{ a -> f(a).ev() }` should fix deriving.